### DISCUSSION TITLE
SwiftUI equivalent components

### DISCUSSION THREAD

**Discussion by jj10133** [AparokshaUI/Adwaita#35] [Improvements]
Firstly @david-swift 🙌🙌 great work... Recently I was looking at swiftui and love writing in declarative way. Was looking on how to make apps for linux until yesterday WWDC lead me here.

I was trying to make a calculator app, however missing Spacer(), more Text() parameters (foreground color...). Please consider me very new to UI and don't have much clue on libadwaita and gtk. 

I willing to help, in writing documentation and code, any group chat or just github?

**Comment by david-swift** [comment]
Thanks for opening the discussion!

> yesterday WWDC lead me here.

How?

> SwiftUI equivalent components

This is definitely an important point. There are some wrappers, type extensions, and aliases making GTK/Libadwaita views easier to use for people familiar with SwiftUI (such as [`Text`](https://aparokshaui.github.io/adwaita-swift/documentation/adwaita/text) for [`Label`](https://aparokshaui.github.io/adwaita-swift/documentation/adwaita/label), [button‘s initializer](https://aparokshaui.github.io/adwaita-swift/documentation/adwaita/button/init(_:handler:)), etc.). Still, a lot of the autogenerated widgets do not have a SwiftUI-like interface. There is still room for many additions making the experience more SwiftUI-like:

- Type aliases
- „Aliases“ for modifiers
- Initializers (many are without any parameters and modifiers should be used for almost everything, while initializers in SwiftUI often have important properties as parameters)

Still, there are differences between GTK/Libadwaita and SwiftUI that make certain concepts difficult to adapt or maybe even not very useful. `Spacer` is such an example. [`valign`](https://aparokshaui.github.io/adwaita-swift/documentation/adwaita/view/valign(_:)), [`halign`](https://aparokshaui.github.io/adwaita-swift/documentation/adwaita/view/halign(_:)), and [`vexpand`](https://aparokshaui.github.io/adwaita-swift/documentation/adwaita/view/vexpand(_:)/), [`hexpand`](https://aparokshaui.github.io/adwaita-swift/documentation/adwaita/view/hexpand(_:)/) are used instead. I think it would be nice to have an article showing a SwiftUI view and its Adwaita counterpart for those cases. 

> any group chat or just github?

Currently just GitHub. Personally, I think it’s great if also questions about the „backend“ are organized in the discussions. Very cool you’re interested in contributing!

**Reply by jj10133** [reply]
Instead of valign or halign could we not use Spacer()?  Like hide underneath

**Comment by jj10133** [comment]
Well past a year I was learning swift and swiftui... I would have to say I love swiftui and swift(something's don't like or may be I still knowledge). So I was following wwdc from 2023, so was waiting for 2024 (this video lead me here -> https://developer.apple.com/videos/play/wwdc2024/10136/)
